tests/python/pants_test/backend/codegen/tasks/test_simple_codegen_task.py

```diff
@@ -0,0 +1,185 @@
+
+    global_dirs = set(get_suffix(target, 'global') for target in targets)
+    isolated_dirs = set(get_suffix(target, 'isolated') for target in targets)
+
+    self.assertEqual(1, len(global_dirs), 'There should only be one global directory suffix!')
+    self.assertEqual(len(targets), len(isolated_dirs),
+                     'There should be exactly one directory suffix per unique target!')
+
+    for target in targets:
+      for strategy in ('global', 'isolated',):
+        suffix = None
+        for i in range(5):
+          curr_suffix = get_suffix(target, strategy)
+          if i > 0:
+            self.assertEqual(suffix, curr_suffix,
+                             'Suffix should not be unstable for the same target and strategy! '
+                             '({target} with {strategy} got {suffix_a} and {suffix_b})'
+                             .format(target=target.id,
+                                     strategy=strategy,
+                                     suffix_a=suffix,
+                                     suffix_b=curr_suffix))
+          suffix = curr_suffix
+
+  def test_execute(self):
+    proto_suffixes = ['a', 'b', 'c',]
+
+    self.add_to_build_file('gen-lib', '\n'.join(dedent('''
+      java_wire_library(name='{suffix}',
+        sources=['org/pantsbuild/example/foo{suffix}.proto'],
+      )
+    ''').format(suffix=suffix) for suffix in proto_suffixes))
+
+    for suffix in proto_suffixes:
+      self.create_file('gen-lib/org/pantsbuild/example/foo{suffix}.proto'.format(suffix=suffix),
+                       dedent('''
+                                package org.pantsbuild.example;
+
+                                message Foo%s {
+                                  optional string bar = 1;
+                                }
+                              ''' % suffix)) # .format() chokes because of the message braces.
+
+    targets = [self.target('gen-lib:{suffix}'.format(suffix=suffix)) for suffix in proto_suffixes]
+
+    test_base = self
+    class TestGen(SimpleCodegenTask):
+      def is_gentarget(self, target):
+        return isinstance(target, JavaWireLibrary)
+      def execute_codegen(self, invalid_targets):
+        if self.codegen_strategy == 'isolated':
+          test_base.assertEqual(1, len(invalid_targets),
+                                'Codegen should execute individually in isolated mode.')
+        elif self.codegen_strategy == 'global':
+          test_base.assertEqual(len(targets), len(invalid_targets),
+                                'Codegen should execute all together in global mode.')
+        for target in invalid_targets:
+          for path in self.sources_generated_by_target(target):
+            message_name = os.path.basename(path).split('.')[0]
+            if not os.path.exists(os.path.join(test_base.build_root, os.path.basename(path))):
+              test_base.create_dir(os.path.basename(path))
+            test_base.create_file(path)
+            f = open(path, 'w')
```

Comment:
use a context manager here:  

```
  with open(path, 'w') as f:
    f.write(...)
```

This will handle cleaning up the close() automatically (even in the case of an exception)